### PR TITLE
feat(huggingface): hugging face integration

### PR DESCRIPTION
**Pull request by ermek-botpress**

### CODE REVIEW

File: integrations/huggingface/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        languageModels.push({
+          id: modelId,
+          name: modelId,
+          description: modelId,
+          tags: [],
+          input: {
+            costPer1MTokens: 0,
+            maxTokens: 200000,
```
Comment (ermek-botpress) [inline integrations/huggingface/src/index.ts:43]: Arbitrary number

File: integrations/huggingface/integration.definition.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import { z, IntegrationDefinition, interfaces } from "@botpress/sdk";
+import { integrationName } from "./package.json";
+
+export default new IntegrationDefinition({
+  name: integrationName,
+  version: "0.0.1",
```
Comment (ptrckbp) [inline integrations/huggingface/integration.definition.ts:6]: this is a humble version number :)

File: integrations/huggingface/src/actions/generateContent.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      type: toolCall.type as "function",
+      function: {
+        name: toolCall.function.name,
+        arguments: toolCall.function.arguments as Record<string, any> | null,
+      },
+    }));
+}
+
+function mapToStopReason(hfFinishReason: string) {
+  switch (hfFinishReason) {
```
Comment (ptrckbp) [inline integrations/huggingface/src/actions/generateContent.ts:26]: this might be more readable with a hashmap (albeit a bit slower)

File: integrations/huggingface/src/actions/generateContent.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      return "other";
+  }
+}
+
+export async function generateContent(
+  input: interfaces.llm.GenerateContentInput,
+  hf: HfInference,
+  models: interfaces.llm.Model[]
+): Promise<interfaces.llm.GenerateContentOutput> {
+  if (!input.model?.id) {
```
Comment (ptrckbp) [inline integrations/huggingface/src/actions/generateContent.ts:45]: this seems a bit odd to me. Why wouldn't the interface provide the model directly?

File: integrations/huggingface/src/actions/generateContent.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  if (input.maxTokens && input.maxTokens > model.output.maxTokens) {
+    throw new InvalidPayloadError(
+      `maxTokens must be less than or equal to ${model.output.maxTokens} for model ID "${modelId}`
+    );
+  }
+
+  if (input.responseFormat === "json_object") {
+    input.systemPrompt =
+      (input.systemPrompt || "") +
+      "\n\nYour response must always be in valid JSON format and expressed as a JSON object.";
```
Comment (ptrckbp) [inline integrations/huggingface/src/actions/generateContent.ts:74]: Can we just have the user add it themselves? Also, are stopTokens part of the huggingface api ? If so we could use "\n}"

Comment (ptrckbp) [reply integrations/huggingface/src/actions/generateContent.ts:74]: If there is no system prompt, your string will start with "\n\n".

File: integrations/huggingface/src/actions/generateContent.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+  const messages = [
+    { role: "system", content: input.systemPrompt },
+    ...input.messages,
+  ];
+
+  const request = {
+    model: modelId,
+    messages,
+    temperature: input.temperature,
```
Comment (ptrckbp) [inline integrations/huggingface/src/actions/generateContent.ts:85]: hugging face temperatures go from 0-100, openai's from 0-1. Does the UI allow the temperature to go above 1 in Botpress?

File: integrations/huggingface/src/actions/generateContent.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        toolCalls: mapToolCalls(choice.message.tool_calls),
+      })),
+      usage: {
+        inputTokens: response.usage.prompt_tokens,
+        outputTokens: response.usage.completion_tokens,
+        //User is charged by HF
+        inputCost: 0,
+        outputCost: 0,
+      },
+    } satisfies interfaces.llm.GenerateContentOutput;
```
Comment (ptrckbp) [inline integrations/huggingface/src/actions/generateContent.ts:110]: I live to see phrases like this!!

File: integrations/huggingface/src/actions/generateContent.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      choices: response.choices.map((choice) => ({
+        index: choice.index,
+        role: "assistant",
+        type: "text",
+        content: choice.message.content,
+        stopReason: mapToStopReason(choice.finish_reason),
+        toolCalls: mapToolCalls(choice.message.tool_calls),
+      })),
+      usage: {
+        inputTokens: response.usage.prompt_tokens,
```
Comment (ptrckbp) [inline integrations/huggingface/src/actions/generateContent.ts:104]: Did you make sure we charge based of inputCost and not inputTokens?

File: integrations/huggingface/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+      return;
+    }
+
+    const modelIds = ctx.configuration.languageModels
+      .split(",")
+      .map((el) => el.trim());
+
+    if (modelIds.length > 10) {
+      logger.forBot().debug(`Too many models provided`);
+      return;
```
Comment (ptrckbp) [inline integrations/huggingface/src/index.ts:19]: why did you add this limit?

File: integrations/huggingface/src/index.ts (hunk 1 above)
Comment (ptrckbp) [reply integrations/huggingface/src/index.ts:43]: can we get the maxTokens it from the api result?

File: integrations/huggingface/src/actions/generateContent.ts (hunk 4 above)
Comment (ptrckbp) [reply integrations/huggingface/src/actions/generateContent.ts:45]: I understand now!

**Review by ptrckbp — APPROVED** [review]
I don't see any blockers. I did leave a few comments, you can have a look. 

otherwise, LGTM!

File: integrations/huggingface/src/index.ts (hunk 1 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/index.ts:43]: Nope, we can't, we only get the input/output tokens

File: integrations/huggingface/src/actions/generateContent.ts (hunk 5 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/actions/generateContent.ts:74]: responseFormat is set by Botpress, not chatbuilders. This piece of code was copied from open ai integration.

File: integrations/huggingface/src/actions/generateContent.ts (hunk 8 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/actions/generateContent.ts:104]: We always charge 0. Because hugging face handles payments. That's why in the "resgister" function all models are saved with price as 0 
```
  languageModels.push({
          id: modelId,
          name: modelId,
          description: modelId,
          tags: [],
          input: {
            costPer1MTokens: 0,
            maxTokens: 200000,
          },
          output: {
            costPer1MTokens: 0,
            maxTokens: 200000,
          },
        });
```

File: integrations/huggingface/src/index.ts (hunk 9 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/index.ts:19]: Idk if there's a state limit, but if there is this will protect it

File: integrations/huggingface/integration.definition.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
+import { z, IntegrationDefinition, interfaces } from "@botpress/sdk";
+import { integrationName } from "./package.json";
+
+export default new IntegrationDefinition({
+  name: integrationName,
```
Comment (franklevasseur) [inline integrations/huggingface/integration.definition.ts:5]: **nitpick:** hardcode integration name here and rm in package.json

File: integrations/huggingface/integration.definition.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  configuration: {
+    schema: z.object({
+      accessToken: z.string(),
+      languageModels: z.string().optional().describe("LLMs to use"),
+    }),
+  },
+  entities: {
+    modelRef: {
+      schema: z.object({
+        id: z.string().describe("Model to use for content generation"),
```
Comment (franklevasseur) [inline integrations/huggingface/integration.definition.ts:18]: **suggestion:** This is where you would list available models normally, but I assume this is the desired behavior since there's so much models on huggingface right ?

File: integrations/huggingface/src/actions/generateContent.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+  }
+}
+
+export async function generateContent(
+  input: interfaces.llm.GenerateContentInput,
+  hf: HfInference,
+  models: interfaces.llm.Model[]
+): Promise<interfaces.llm.GenerateContentOutput> {
+  if (!input.model?.id) {
+    throw new InvalidPayloadError("Model ID not provided");
```
Comment (franklevasseur) [inline integrations/huggingface/src/actions/generateContent.ts:46]: **note:** I'm not sure of what will be the behavior of throwing this error in the bots

File: integrations/huggingface/src/actions/generateContent.ts (hunk 6 above)
Comment (franklevasseur) [reply integrations/huggingface/src/actions/generateContent.ts:85]: @ptrckbp if this is true just scale it to 100 like:
```ts
const huggingFaceTemp = openaiTemp * 100
```

File: integrations/huggingface/src/index.ts (hunk 9 above)
Comment (franklevasseur) [reply integrations/huggingface/src/index.ts:19]: @ermek-botpress its a good practice in general, lets bump it if needed. However, I think it would be better to throw a `RuntimeError` here so that the registering step fails louder

File: integrations/huggingface/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+    await client.setState({
+      type: "integration",
+      name: "availableModels",
+      id: ctx.integrationId,
+      payload: { languageModels },
+    });
+  },
+  unregister: async ({ client, ctx }) => {
+    await client.setState({
```
Comment (franklevasseur) [inline integrations/huggingface/src/index.ts:69]: **chore:** I think if you pass a `null` payload it will delete the state

File: integrations/huggingface/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+    if (modelIds.length > 10) {
+      logger.forBot().debug(`Too many models provided`);
+      return;
+    }
+    const languageModels: interfaces.llm.Model[] = [];
+    for await (const id of modelIds) {
+      try {
+        const { data }: { data: { modelId: string; pipeline_tag: string } } =
+          await axios.get(`https://huggingface.co/api/models/${id}`);
```
Comment (franklevasseur) [inline integrations/huggingface/src/index.ts:25]: **nitpick:** I would personally extract this call in a dedicated function and a dedicated file. The reason is; this code violates the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle).

**Review by franklevasseur — APPROVED** [review]
good work

File: integrations/huggingface/src/actions/generateContent.ts (hunk 3 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/actions/generateContent.ts:26]: I went with switch case because it successfully infers return types as string literals, opposed to map which infers it as a generic string type.Screenshot 2024-08-29 at 11 50 57 AM
Screenshot 2024-08-29 at 11 51 45 AM

File: integrations/huggingface/src/actions/generateContent.ts (hunk 6 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/actions/generateContent.ts:85]: Anything above 1.5 turns out to be gibberish, so we'll keep it below 1

File: integrations/huggingface/integration.definition.ts (hunk 11 above)
Comment (ermek-botpress) [reply integrations/huggingface/integration.definition.ts:18]: Exactly, we add only llms, that user adds in the integration configuration as such: https://i.imgur.com/2gx0wnB.png

File: integrations/huggingface/src/actions/generateContent.ts (hunk 12 above)
Comment (ermek-botpress) [reply integrations/huggingface/src/actions/generateContent.ts:46]: It is identical to sdk.RuntimeError from what I see
